[Docs] Add node ESM behaviour in import/extensions (#2748)
import-js/eslint-plugin-import: docs/rules/extensions.md

<!-- end auto-generated rule header -->

> Some file resolve algorithms allow you to omit the file extension within the import source path. For example the `node` CommonJS resolver can resolve `./foo/bar` to the absolute path `/User/someone/foo/bar.js` because the `.js` extension is resolved automatically by default. Depending on the resolver you can configure more extensions to get resolved automatically.

Reviewer: it's a node resolver, not a CJS resolver - it's just that it doesn't yet support "exports" or ESM.

Author: To confirm we're on the same page here. My understanding is that `node` in this context refers specifically to NodeJS, is that correct?

NodeJS documents two separate module loaders, one for CommonJS and one for ECMAScript. The ESM loader explicitly says:

> _It does no extension searching. A file extension must be provided when the specifier is a relative or absolute file URL._

https://nodejs.org/api/packages.html#modules-loaders

Reviewer: yes, but "node" describes many node versions over a long period of time, it's not inherently "node latest". It's a node resolver because it matches node's resolution *at one point* and because it intends to match the latest version eventually.

Yes, the ESM loader does no extension searching, but any time it's resolving something in a package's "exports" field, the specifier need not (and often should not!) have an extension. It's only on relative files that users are forced to suffer the pain of specifying extensions.

Author: So this proposed change is just updating the `eslint-plugin-import` documentation to be correct now that NodeJS has more that one resolver, right? NodeJS has had more than one resolver since 2017 and referring to them as the same for this scenario is misleading.

>Yes, the ESM loader does no extension searching, but any time it's resolving something in a package's "exports" field, the specifier need not (and often should not!) have an extension. It's only on relative files that users are forced to suffer the pain of specifying extensions.

Given the rule:

```json
{
  "import/extensions": ["error", "always"],
}
```

This code is correct:

```js
import assert from "assert";

import {aFunc, bFunc} from "../local.js";
```

So `import/extensions` already correctly only enforces extensions on relative files and leaves packages untouched.

Reviewer: That's not accurate - node's ESM resolver became unflagged in 2020, not before, and it wasn't until April 2021 that the last non-ESM LTS line went EOL (i was in the node modules working group; i'm quite familiar with the timeline)

Author: Then you'll no doubt also know that NodeJS's ESM resolver was available behind a flag and/or using `.mjs` since [v8 back in 2017](https://nodejs.org/docs/latest-v8.x/api/esm.html). That said, I'm not interested in debating timelines.

I found the `import/extensions` documentation in it's current state to be confusing as it has examples using the `import` syntax and specifically discusses "`node`" as not requiring a file extension for local files which as far as I can tell has never been the case. Because I wasted time trying this rule out and having to revert those changes - and had seen another user already raised this before (#2043) - I thought it might be useful for other users to clarify this.

Feel free to suggest alternative wording, I am happy to rephrase my changes. But I can't tell what you're looking for from your feedback so far.

Reviewer: Behind a flag doesn't count :-)

Let me give it a fresh review and see what I can suggest.

> Some file resolve algorithms allow you to omit the file extension within the import source path. For example the `node` CommonJS resolver can resolve `./foo/bar` to the absolute path `/User/someone/foo/bar.js` because the `.js` extension is resolved automatically by default. Depending on the resolver you can configure more extensions to get resolved automatically.

Reviewer: ```suggestion
Some file resolve algorithms allow you to omit the file extension within the import source path. For example the `node` resolver (which does not yet support ESM/`import`) can resolve `./foo/bar` to the absolute path `/User/someone/foo/bar.js` because the `.js` extension is resolved automatically by default in CJS. Depending on the resolver you can configure more extensions to get resolved automatically.
```

## When Not To Use It

If you are not concerned about a consistent usage of file extension.

> You should not configure this rule to be "never" if you are using ECMAScript modules directly in `node`.

Reviewer: I don't think this is exactly accurate advice - when the node resolver does eventually support "exports" and ESM, "never" would not apply to anything that required the extension, but would correctly forbid it for anything where the extension was optional.

Author: The documentation for NodeJS has marked _Modules: ECMAScript modules_ as _stable_ since v15.3 - although other parts of the ESM spec are still marked experimental. They've explicitly required the `.js` extension for the ECMAScript loader [since v12](https://nodejs.org/docs/latest-v12.x/api/esm.html#esm_mandatory_file_extensions) (back in 2019).

Do you know that they're actually planning future changes to this part of the `import` behaviour specifically? Having something about the current behaviour in NodeJS would be useful!

> You should not configure this rule to be "never" if you are using ECMAScript modules directly in `node`.

Reviewer: ```suggestion
In the future, when this rule supports native node ESM resolution (and the plugin is configured to use native, instead of transpiled, ESM, a config option that is not yet available), setting this to `always` will have no effect.
```

> > A file extension must be provided when using the import keyword to resolve relative or absolute specifiers.
> > Directory indexes (e.g. `'./startup/index.js'`) must also be fully specified.
> 
> See [esm#mandatory-file-extensions](https://nodejs.org/api/esm.html#mandatory-file-extensions) for more details.

Reviewer: ```suggestion
```

## When Not To Use It

If you are not concerned about a consistent usage of file extension.

> In the future, when this rule supports native node ESM resolution (and the plugin is configured to use native, instead of transpiled, ESM, a config option that is not yet available), setting this to `always` will have no effect.

Author: On second thought, this sentence clause has quite a few commas. What do you think of:

```suggestion
In the future, when this rule supports native node ESM resolution, and the plugin is configured to use native rather than transpiled ESM (a config option that is not yet available) - setting this to `always` will have no effect.
```

Reviewer: sure, that's fine
